### PR TITLE
Fix geo_shape centroid calculation for downgraded shapes

### LINKED ISSUE #52303: geo-shape-doc-values centroid calculation should handle polygons as lines/points and lines as points
- Polygons with no area should not be treated as polygons — they should be treated as lines.
- Lines with no length should not be treated as lines — they should be treated as points.

**possible solution A**

This can be achieved by updating the `DimensionalShapeType` of the geometry for such geometries. The downside of this is that then we would love information about what type of geometry the field value was.

**possible solution B**

The other option is to keep the original `DimensionalShapeType`, and instead have a place-holder weight that is close to `0` for such geometries.

relates to #50834

### PR DESCRIPTION
**Pull request by talevy**
This commit modifies the centroid-calculator/dimensional-shape-type
to properly support the instances of polygons that have no area
and lines that have no length. Beforehand N/A were returned for the
centroid values, but it is best to downcast the shape type to
the appropriate type.

Closes #52303.

### CODE REVIEW

**Comment by talevy** [comment]
run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/common/geo/CentroidCalculator.java [hunk 1]
```diff
@@ -31,29 +31,45 @@
...
 public class CentroidCalculator {
-    private double compX;
-    private double compY;
-    private double sumX;
-    private double sumY;
-    private double sumWeight;
+    CompensatedSum compSumX;
+    CompensatedSum compSumY;
+    CompensatedSum compSumWeight;
+
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/common/geo/CentroidCalculator.java]: Shouldn't we make these private as we have getters for them?

**Review by iverase — APPROVED** [review]
lgtm. Just left a comment but feel free to ignore it :)

**Comment by talevy** [comment]
test this please

File: server/src/main/java/org/elasticsearch/common/geo/CentroidCalculator.java (hunk 1 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/CentroidCalculator.java]: I've removed the getters! thanks

**Comment by talevy** [comment]
@elasticmachine test this please

**Comment by talevy** [comment]
run elasticsearch-ci/2

**Comment by talevy** [comment]
run elasticsearch-ci/packaging-sample-matrix-unix

**Comment by talevy** [comment]
run elasticsearch-ci/1

**Pull request merged**